File: CrossTimeDsp/Dsp/FirstOrderFilterDouble.cs
using System;
using System.Diagnostics;

namespace CrossTimeDsp.Dsp
{
    /// <summary>
    /// Direct form 1 implementation of a first order filter.
    /// </summary>
    internal class FirstOrderFilterDouble : IFilter<double>
    {
        private double a1;
        private double b0;
        private double b1;
        private int channels;
        private double[] x1;
        private double[] y1;

        public FirstOrderFilterDouble(double b0, double b1, double a0, double a1, int channels)
        {
            this.a1 = a1 / a0;
            this.b0 = b0 / a0;
            this.b1 = b1 / a0;

            this.channels = channels;
            this.x1 = new double[channels];
            this.y1 = new double[channels];
        }

        public void Filter(double[] block, int offset)
        {
            int maxSample = offset + Constant.FilterBlockSizeInDoubles;
            if (this.channels == 2)
            {
                Debug.Assert(offset % 2 == 0, String.Format("Offset {0} is not aligned to stereo block.", offset));
                for (int sample = offset; sample < maxSample; sample += 2)
                {
                    double value0 = block[sample];
                    double accumulator0 = this.b0 * value0 + this.b1 * this.x1[0] - this.a1 * this.y1[0];
                    this.x1[0] = value0;
                    this.y1[0] = accumulator0;
                    block[sample] = accumulator0;

                    double value1 = block[sample + 1];
                    double accumulator1 = this.b0 * value1 + this.b1 * this.x1[1] - this.a1 * this.y1[1];
                    this.x1[1] = value1;
                    this.y1[1] = accumulator1;
                    block[sample + 1] = accumulator1;
                }
            }
            else
            {
                for (int sample = offset; sample < maxSample; ++sample)
                {
                    int channel = sample % this.channels;
                    double value = block[sample];
                    double accumulator = this.b0 * value + this.b1 * this.x1[channel] - this.a1 * this.y1[channel];
                    this.x1[channel] = value;
                    this.y1[channel] = accumulator;
                    block[sample] = accumulator;
                }
            }
        }

        public void FilterReverse(double[] block, int offset)
        {
            int maxSample = offset + Constant.FilterBlockSizeInDoubles;
            if (this.channels == 2)
            {
                Debug.Assert(offset % 2 == 0, String.Format("Offset {0} is not aligned to stereo block.", offset));
                for (int sample = maxSample - 2; sample >= offset; sample -= 2)
                {
                    double value0 = block[sample];
                    double accumulator0 = this.b0 * value0 + this.b1 * this.x1[0] - this.a1 * this.y1[0];
                    this.x1[0] = value0;
                    this.y1[0] = accumulator0;
                    block[sample] = accumulator0;

                    double value1 = block[sample + 1];
                    double accumulator1 = this.b0 * value1 + this.b1 * this.x1[1] - this.a1 * this.y1[1];
                    this.x1[1] = value1;
                    this.y1[1] = accumulator1;
                    block[sample + 1] = accumulator1;
                }
            }
            else
            {
                for (int sample = maxSample - 1; sample >= offset; --sample)
                {
                    int channel = sample % this.channels;
                    double value = block[sample];
                    double accumulator = this.b0 * value + this.b1 * this.x1[channel] - this.a1 * this.y1[channel];
                    this.x1[channel] = value;
                    this.y1[channel] = accumulator;
                    block[sample] = accumulator;
                }
            }
        }
    }
}

File: Native/FirstOrderFilterDouble.cpp
#include "stdafx.h"
#include <new>
#include "FirstOrderFilterDouble.h"

namespace CrossTimeDsp::Dsp
{
	#pragma unmanaged
	FirstOrderFilterDouble::FirstOrderFilterDouble(double b0, double b1, double a0, double a1, __int32 channels)
	{
		this->channels = channels;
		if (channels == 2)
		{
			this->a1_2 = _mm_set1_pd(a1 / a0);
			this->b0_2 = _mm_set1_pd(b0 / a0);
			this->b1_2 = _mm_set1_pd(b1 / a0);

			this->x1_2 = _mm_setzero_pd();
			this->y1_2 = _mm_setzero_pd();
		}
		else
		{
			this->a1_n = a1 / a0;
			this->b0_n = b0 / a0;
			this->b1_n = b1 / a0;

			this->x1_n = new double[channels];
			this->y1_n = new double[channels];
			for (int channel = 0; channel < channels; ++channel)
			{
				this->x1_n[channel] = 0.0;
				this->y1_n[channel] = 0.0;
			}
		}
	}

	FirstOrderFilterDouble::~FirstOrderFilterDouble()
	{
		delete this->x1_n;
		delete this->y1_n;
	}

	void FirstOrderFilterDouble::Filter(double* block, __int32 offset)
	{
		int maxSample = offset + Constant::FilterBlockSizeInDoubles;
		if (this->channels == 2)
		{
			for (int sample = offset; sample < maxSample; sample += 2)
			{
				__m128d values = _mm_loadu_pd(block + sample);
				__m128d accumulator = _mm_mul_pd(this->b0_2, values);
						accumulator = _mm_add_pd(accumulator, _mm_mul_pd(this->b1_2, this->x1_2));
						accumulator = _mm_sub_pd(accumulator, _mm_mul_pd(this->a1_2, this->y1_2));
				this->x1_2 = values;
				this->y1_2 = accumulator;
				_mm_storeu_pd(block + sample, accumulator);
			}
		}
		else
		{
			for (int sample = offset; sample < maxSample; ++sample)
			{
				int channel = sample % this->channels;
				double value = block[sample];
				double accumulator = this->b0_n * value + this->b1_n * this->x1_n[channel] - this->a1_n * this->y1_n[channel];
				this->x1_n[channel] = value;
				this->y1_n[channel] = accumulator;
				block[sample] = accumulator;
			}
		}
	}

	void FirstOrderFilterDouble::FilterReverse(double* block, __int32 offset)
	{
		__int32 maxSample = offset + Constant::FilterBlockSizeInDoubles;
		if (this->channels == 2)
		{
			for (int sample = maxSample - 2; sample >= offset; sample -= 2)
			{
				__m128d values = _mm_loadu_pd(block + sample);
				__m128d accumulator = _mm_mul_pd(this->b0_2, values);
						accumulator = _mm_add_pd(accumulator, _mm_mul_pd(this->b1_2, this->x1_2));
						accumulator = _mm_sub_pd(accumulator, _mm_mul_pd(this->a1_2, this->y1_2));
				this->x1_2 = values;
				this->y1_2 = accumulator;
				_mm_storeu_pd(block + sample, accumulator);
			}
		}
		else
		{
			for (int sample = maxSample - 1; sample >= offset; --sample)
			{
				int channel = sample % this->channels;
				double value = block[sample];
				double accumulator = this->b0_n * value + this->b1_n * this->x1_n[channel] - this->a1_n * this->y1_n[channel];
				this->x1_n[channel] = value;
				this->y1_n[channel] = accumulator;
				block[sample] = accumulator;
			}
		}
	}
}